### PR TITLE
Promote `ServiceReference` to stable

### PR DESCRIPTION
**Pull request by abstratt**
Fixes: #30858

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
@bot-gradle test RFN

File: platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc [hunk 1]
```diff
@@ -73,41 +73,30 @@ Note that you should *not* implement the link:{javadocPath}/org/gradle/api/servi
 A build service implementation must be thread-safe, as it will potentially be used by multiple tasks concurrently.
 
 [[using_a_build_service_from_a_task]]
```
Comment (abstratt) [inline platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc:75]: @lkasso I felt the existing documentation had some unnecessary duplication. I merged the sections about registering and consuming build services, and promoted `@ServiceReference` as preferred over `@Internal` properties.

**Comment by abstratt** [comment]
@bot-gradle test BD

**Comment by abstratt** [comment]
Release notes (which links to the documentation that was updated): 

https://builds.gradle.org/repository/download/Gradle_Master_Check_BuildDistributions/90469529:id/distributions/gradle-8.12-docs.zip!/gradle-8.12-20241018205650%2B0000/docs/release-notes.html#promoted-features

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java [hunk 2]
```diff
@@ -841,9 +841,13 @@ public String determineName(Task task) {
     /**
      * Registers a {@link BuildService} that is used by this task so
      * {@link BuildServiceRegistration#getMaxParallelUsages() its constraint on parallel execution} can be honored.
+     * <p>
+     * This is not necessary for task properties declared as {@link org.gradle.api.services.ServiceReference}s.
+     * </p>
```
Comment (alllex) [inline subprojects/core-api/src/main/java/org/gradle/api/Task.java:846]: 💅  We are not using closing paragraph tags in javadoc as per our javadoc style guide

File: subprojects/core-api/src/main/java/org/gradle/api/services/BuildServiceRegistration.java [hunk 3]
```diff
@@ -38,10 +38,11 @@ public interface BuildServiceRegistration<T extends BuildService<P>, P extends B
      * When this property has no value defined, then any number of tasks may use this service in parallel. This is the default.
      *
      * <p>
-     * IMPORTANT: the build service must be explicitly registered with every using task via {@link org.gradle.api.Task#usesService(Provider) Task#usesService}
+     * IMPORTANT: the build service must be consumed via a {@link ServiceReference} property, or explicitly registered with every using task via {@link org.gradle.api.Task#usesService(Provider) Task#usesService}
```
Comment (alllex) [inline subprojects/core-api/src/main/java/org/gradle/api/services/BuildServiceRegistration.java:41]: 💅  Splitting the line in two for better readability in unrendered mode?

File: subprojects/core-api/src/main/java/org/gradle/api/services/BuildServiceSpec.java [hunk 4]
```diff
@@ -41,10 +42,11 @@ public interface BuildServiceSpec<P extends BuildServiceParameters> {
      * When this property has no value defined, then any number of tasks may use this service in parallel. This is the default.
      *
      * <p>
-     * IMPORTANT: the build service must be explicitly registered with every using task via {@link org.gradle.api.Task#usesService(Provider) Task#usesService}
+     * IMPORTANT: the build service must be consumed via a {@link ServiceReference} property, or explicitly registered with every using task via {@link org.gradle.api.Task#usesService(Provider) Task#usesService}
```
Comment (alllex) [inline subprojects/core-api/src/main/java/org/gradle/api/services/BuildServiceSpec.java:45]: 💅  Also a very long line

File: .idea/misc.xml
Comment (alllex) [inline .idea/misc.xml:1]: ❌  These changes should not be part of this PR

Comment (abstratt) [reply .idea/misc.xml:1]: I wanted to merge that change though, it was not accidental. Are you suggesting I opened a separate PR for it?

Comment (alllex) [reply .idea/misc.xml:1]: Yes, I'd also share this with the larger team to review. Personally, I don't know what effects will this change have.
It seems to be too independent of a change.

**Review by mlopatkin — CHANGES_REQUESTED** [review]
I really like the updated documentation! I've left one question about the suggested replacement, which, I think, we should resolve before merging.

File: subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProviderNagger.java [hunk 5]
```diff
@@ -51,7 +51,7 @@ private static boolean isServiceRequiredBy(TaskInternal task, BuildServiceProvid
     private static void nagAboutUndeclaredUsageOf(BuildServiceProvider<?, ?> provider, TaskInternal task) {
         deprecateBehaviour(undeclaredBuildServiceUsage(provider, task))
             .withProblemIdDisplayName("Build Service " + provider.getName()+ " undeclared usage")
-            .withAdvice("Declare the association between the task and the build service using 'Task#usesService'.")
+            .withAdvice("Declare the association between the task by declaring the consuming property as a '@ServiceReference'.")
```
[the remark this answers is not in the document]
Comment (abstratt) [reply subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProviderNagger.java:54]: Indeed, if the task cannot be changed to declare a property to reference the second service, `usesService` allows one to address the issue without requiring a modification to the task definition itself. 

I considered that and decided to omit that advice since it is more of an edge case and we don't want to encourage folks that don't need to use `usesService` to use it.

That would be a workaround for https://github.com/gradle/gradle/issues/24780 btw.

I will look into updating the documentation so it mentions `usesService` as a workaround regarding indirect service usage. If that does not work out, I will update the message above.

File: subprojects/core-api/src/main/java/org/gradle/api/Task.java (hunk 2 above)
Comment (abstratt) [reply subprojects/core-api/src/main/java/org/gradle/api/Task.java:846]: It is quite widespread on this file though... I kinda feel local consistency is more important.

Comment (abstratt) [reply subprojects/core-api/src/main/java/org/gradle/api/Task.java:846]: Will drop the closing tag I introduced in the other source files

File: platforms/documentation/docs/src/docs/release/notes.md [hunk 6]
```diff
@@ -94,9 +94,10 @@ See the User Manual section on the “[Feature Lifecycle](userguide/feature_life
 
 The following are the features that have been promoted in this Gradle release.
 
-<!--
-### Example promoted
--->
+### Service reference properties are now stable
+
+Service references are task properties meant for easier consumption of shared build services.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:99]:
```suggestion
Service references are task properties meant for easier consumption of [shared build services](userguide/build_services.html#sec:service_references).
```

File: platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc [hunk 7]
```diff
@@ -73,41 +73,30 @@ Note that you should *not* implement the link:{javadocPath}/org/gradle/api/servi
...
-3. Assign a shared build service provider to the property (optional, when using `@ServiceReference(<serviceName>)`).
-4. Declare the association between the task and the service so Gradle can properly honor the build service lifecycle and its usage constraints (also optional when using `@ServiceReference`).
-
-Note that using a service with any other annotation is currently not supported.
-For example, it is currently impossible to mark a service as an input to a task.
+== Registering a build service and connecting it to a task
 
-=== Annotating a shared build service property with `@Internal`
+To create a build service, you register the service instance using the link:{javadocPath}/org/gradle/api/services/BuildServiceRegistry.html#registerIfAbsent-java.lang.String-java.lang.Class-org.gradle.api.Action-[BuildServiceRegistry.registerIfAbsent()] method.
+Registering the service does not create the service instance.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc:79]:
```suggestion

Registering the service does not create the service instance.
```

File: platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc [hunk 8]
```diff
@@ -184,8 +171,9 @@ When this property has no value, which is the default, Gradle does not constrain
 When this property has a value > 0, Gradle will allow no more than the specified number of tasks to use the service concurrently.
 
 IMPORTANT: When the consuming task property is annotated with `@Internal`, for the constraint to take effect, the build service *must* be registered with the consuming task via
-link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:usesService(org.gradle.api.provider.Provider)[Task.usesService(Provider<? extends BuildService<?>>)].
-This is not necessary if, instead, the consuming property is annotated with `@ServiceReference`.
+link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:usesService(org.gradle.api.provider.Provider)[Task.usesService(Provider<? extends BuildService<?>>)]. NOTE: at this time, Gradle cannot discover indirect usage of services (for instance, if an additional service is used only by a service that the task uses directly).
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc:174]:
```suggestion
link:{groovyDslPath}/org.gradle.api.Task.html#org.gradle.api.Task:usesService(org.gradle.api.provider.Provider)[Task.usesService(Provider<? extends BuildService<?>>)]. 
At this time, Gradle cannot discover indirect service usage (for instance, if an additional service is used only by a service that the task uses directly).
```

File: platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc (hunk 1 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc:75]: I like this update a lot.

File: platforms/documentation/docs/src/docs/release/notes.md [hunk 9]
```diff
@@ -94,9 +94,10 @@ See the User Manual section on the “[Feature Lifecycle](userguide/feature_life
...
+See documentation on [shared build services](userguide/build_services.html#sec:service_references) for more information.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:100]:
```suggestion
```

File: platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc (hunk 8 above)
Comment (abstratt) [reply platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc:174]: I tried something like that, but this would break formatting - lines 174/175 would no longer be part of the "IMPORTANT" block, and they are additions related to the content in that block. I considered having two IMPORTANT blocks, one after the other, and it felt weird to me. 

Any alternative suggestions, @lkasso?

**Comment by abstratt** [comment]
@bot-gradle test BD

**Review by mlopatkin — APPROVED** [review]

File: platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc (hunk 8 above)
Comment (lkasso) [reply platforms/documentation/docs/src/docs/userguide/authoring-builds/tasks/build_services.adoc]: Two solutions:
1. Feel free to keep it on the same line but just change it to:  `NOTE: at this` -> `Note that at this`
2. You can also use a `+` to make the formating work. `NOTE: at this` -> 
```
Task.usesService(Provider<? extends BuildService<?>>)].  
+
Note that at this time,
```
